[pt][quant] Enable the dynamic quant LSTM unit test by removing hypothesis

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33594 [pt][quant] Enable the dynamic quant LSTM unit test by removing hypothesis**

Check more discussions in https://github.com/pytorch/pytorch/issues/32644.

Previously, after landing https://github.com/pytorch/pytorch/pull/32757, we found the flaky tests on `test_quantized_rnn` in some specific CI environment, where we cannot reproduce in our devserver.

The error is reported as it failed the `scale > 0` check in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L69 .

- Step 1: we disabled the unit tests in https://github.com/pytorch/pytorch/pull/32742

- Step 2: we investigated the FBGEMM `ChooseQuantizationParams` routine and added assert for min, max, qmin, qmax, but we didn't find the abnormal numbers :
https://github.com/pytorch/pytorch/pull/32739

scale is calculated as the following:
float scale = (static_cast<double>(max) - min) / (qmax - qmin);

If max is larger than min, and qmax is larger than qmin, then scale should be always positive.

- Step 3: we tried to print the intermediate result of x_min and x_max in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L59-L63. Check https://github.com/pytorch/pytorch/pull/33167 for the places we add the prints.

One interesting test case caused the failure:
```
Jan 29 23:46:57 x_min: -nan, x_max: -nan
Jan 29 23:46:57 input: -nan 0.195275 -0.380797 -0.510193 -0.662963 0.761594 0.923773 -nan -0.123308 -0.380797 -0.761594 -0.603171 0.906278 0.000028 -nan 0.462941 -0.380797 -0.651336 0.717926 0.761594 0.460751 -nan 0.945544 -0.380797 -0.334840 0.831535 0.761594 0.793159 -nan -0.752703 -0.380797 -0.761594 0.809177 0.761594 0.955718 -nan -0.717292 -0.380797 -0.650404 0.790869 0.761594 0.812008 -nan 0.917088 -0.380797 -0.761594 0.829536 0.761594 0.000000 -nan 0.855218 -0.380796 -0.036113 0.815578 0.761594 0.942870 -nan -0.675926 -0.380797 -0.755426 -0.658459 0.000000 0.000000 -nan 0.490974 -0.380797 -0.725917 -0.539679 0.761594 0.836110 -nan -0.714314 -0.380797 -0.761594 -0.738534 0.761594 0.791485 -nan 0.345541 -0.380797 -0.670953 -0.609272 0.962786 0.834005 -nan -0.583866 -0.380797 -0.761594 -0.620485 0.750284 0.854029 -nan -0.661655 -0.380797 -0.282387 0.863691 0.761593 0.956279 -nan -0.667277 -0.380797 -0.730052 -0.732839 0.761594 0.002185 -nan 0.380570 -0.380797 -0.453875 -0.523069 0.834210 0.792132
```

More detailed test results are here:
https://circleci.com/gh/pytorch/pytorch/4374802?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Basically, x_min and x_max are both -nan, which failed the check for x_min <= x_max. Actually if any operands for a comparison is nan (e.g., nan < 3, nan < nan, nan <= nan), the result for the comparison will always be False. The culprit should be this NaN value.

The issue is where we generated such NaN value?

- Step 4:
We analyzed that NaN can possibly come from two sources:

    - Unit test: when when we initialize the `torch.Tensor` or `torch.rand` somewhere, it generates NaN number;

    - Somewhere in `aten/src/ATen/native/RNN.cpp` or `torch/nn/quantized/dynamic/modules/rnn.py`, we used some illegal instructions.

As there is no such issue before https://github.com/pytorch/pytorch/pull/32757, we suspect some changes in that PR triggered this issue.

In https://github.com/pytorch/pytorch/pull/33167, we avoided using hypothesis by removing `@given(qengine=st.sampled_from(("qnnpack", "fbgemm"))):` and all tests pass.

It looks like that the hypothesis added in https://github.com/pytorch/pytorch/pull/32757 triggered the generation of NaN in some input tensors.


- Step 5:

We still need to find the root cause for NaN, either it is the problem of hypothesis itself, or hypothesis generated more random cases which helped triggered some NaN numbers in `RNN.cpp` or `rnn.py`.

Differential Revision: [D20016038](https://our.internmc.facebook.com/intern/diff/D20016038/)

jianyuh committed:
[pt][quant] Enable the dynamic quant LSTM unit test by removing hypothesis

Check more discussions in https://github.com/pytorch/pytorch/issues/32644.

Previously, after landing https://github.com/pytorch/pytorch/pull/32757, we found the flaky tests on `test_quantized_rnn` in some specific CI environment, where we cannot reproduce in our devserver.

The error is reported as it failed the `scale > 0` check in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L69 .

- Step 1: we disabled the unit tests in https://github.com/pytorch/pytorch/pull/32742

- Step 2: we investigated the FBGEMM `ChooseQuantizationParams` routine and added assert for min, max, qmin, qmax, but we didn't find the abnormal numbers :
https://github.com/pytorch/pytorch/pull/32739

scale is calculated as the following:
float scale = (static_cast<double>(max) - min) / (qmax - qmin);

If max is larger than min, and qmax is larger than qmin, then scale should be always positive.

- Step 3: we tried to print the intermediate result of x_min and x_max in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L59-L63. Check https://github.com/pytorch/pytorch/pull/33167 for the places we add the prints.

One interesting test case caused the failure:
```
Jan 29 23:46:57 x_min: -nan, x_max: -nan
Jan 29 23:46:57 input: -nan 0.195275 -0.380797 -0.510193 -0.662963 0.761594 0.923773 -nan -0.123308 -0.380797 -0.761594 -0.603171 0.906278 0.000028 -nan 0.462941 -0.380797 -0.651336 0.717926 0.761594 0.460751 -nan 0.945544 -0.380797 -0.334840 0.831535 0.761594 0.793159 -nan -0.752703 -0.380797 -0.761594 0.809177 0.761594 0.955718 -nan -0.717292 -0.380797 -0.650404 0.790869 0.761594 0.812008 -nan 0.917088 -0.380797 -0.761594 0.829536 0.761594 0.000000 -nan 0.855218 -0.380796 -0.036113 0.815578 0.761594 0.942870 -nan -0.675926 -0.380797 -0.755426 -0.658459 0.000000 0.000000 -nan 0.490974 -0.380797 -0.725917 -0.539679 0.761594 0.836110 -nan -0.714314 -0.380797 -0.761594 -0.738534 0.761594 0.791485 -nan 0.345541 -0.380797 -0.670953 -0.609272 0.962786 0.834005 -nan -0.583866 -0.380797 -0.761594 -0.620485 0.750284 0.854029 -nan -0.661655 -0.380797 -0.282387 0.863691 0.761593 0.956279 -nan -0.667277 -0.380797 -0.730052 -0.732839 0.761594 0.002185 -nan 0.380570 -0.380797 -0.453875 -0.523069 0.834210 0.792132
```

More detailed test results are here:
https://circleci.com/gh/pytorch/pytorch/4374802?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Basically, x_min and x_max are both -nan, which failed the check for x_min <= x_max. Actually if any operands for a comparison is nan (e.g., nan < 3, nan < nan, nan <= nan), the result for the comparison will always be False. The culprit should be this NaN value.

The issue is where we generated such NaN value?

- Step 4:
We analyzed that NaN can possibly come from two sources:

    - Unit test: when when we initialize the `torch.Tensor` or `torch.rand` somewhere, it generates NaN number;

    - Somewhere in `aten/src/ATen/native/RNN.cpp` or `torch/nn/quantized/dynamic/modules/rnn.py`, we used some illegal instructions.

As there is no such issue before https://github.com/pytorch/pytorch/pull/32757, we suspect some changes in that PR triggered this issue.

In https://github.com/pytorch/pytorch/pull/33167, we avoided using hypothesis by removing `@given(qengine=st.sampled_from(("qnnpack", "fbgemm"))):` and all tests pass.

It looks like that the hypothesis added in https://github.com/pytorch/pytorch/pull/32757 triggered the generation of NaN in some input tensors.


- Step 5:

We still need to find the root cause for NaN, either it is the problem of hypothesis itself, or hypothesis generated more random cases which helped triggered some NaN numbers in `RNN.cpp` or `rnn.py`.

Differential Revision: [D20016038](https://our.internmc.facebook.com/intern/diff/D20016038/)

[ghstack-poisoned]

jianyuh commented:
@supriyar : The new unit tests with hypothesis show some failure with QNNPACK in https://circleci.com/gh/pytorch/pytorch/4557111?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link:
```
Feb 21 02:16:17 ...../var/lib/jenkins/workspace/caffe2/utils/threadpool/pthreadpool.cc:332:13: runtime error: call to function qnnpack::compute_q8gemm_dq(qnnpack::q8gemm_dq_context const*, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long) through pointer to incorrect function type 'void (*)(void *, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long)'
Feb 21 02:16:17 (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x12a8ebb0): note: qnnpack::compute_q8gemm_dq(qnnpack::q8gemm_dq_context const*, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long) defined here
Feb 21 02:16:17     #0 0x7fd4da18207f in pthreadpool_compute_4d_tiled (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xeb7507f)
Feb 21 02:16:17     #1 0x7fd4de09ba29 in qnnpack::qnnpackLinearDynamic(unsigned long, unsigned long, unsigned long, unsigned char, float, unsigned char, float, unsigned char const*, unsigned long, void*, float const*, float*, unsigned long, pthreadpool*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x12a8ea29)
Feb 21 02:16:17     #2 0x7fd4d722d5c4 in at::native::(anonymous namespace)::QLinearDynamicInt8<false>::operator()(at::Tensor, at::Tensor) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xbc205c4)
Feb 21 02:16:17     #3 0x7fd4d722aa2c in c10::detail::make_boxed_from_unboxed_functor<at::native::(anonymous namespace)::QLinearDynamicInt8<false>, false, void>::call(c10::OperatorKernel*, c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xbc1da2c)
Feb 21 02:16:17     #4 0x7fd4d657a0aa in c10::KernelFunction::callBoxed(c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xaf6d0aa)
Feb 21 02:16:17     #5 0x7fd4d65793a4 in (anonymous namespace)::variable_fallback_kernel(c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xaf6c3a4)
Feb 21 02:16:17     #6 0x7fd4d657a0aa in c10::KernelFunction::callBoxed(c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xaf6d0aa)
Feb 21 02:16:17     #7 0x7fd4d6ffb263 in std::vector<c10::IValue, std::allocator<c10::IValue> > callOp<at::Tensor, at::Tensor>(c10::OperatorHandle const&, at::Tensor, at::Tensor) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xb9ee263)
Feb 21 02:16:17     #8 0x7fd4d6ffafc8 in std::vector<c10::IValue, std::allocator<c10::IValue> > callOp<at::Tensor, at::Tensor>(char const*, char const*, at::Tensor, at::Tensor) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xb9edfc8)
Feb 21 02:16:17     #9 0x7fd4d6fb3d65 in at::native::(anonymous namespace)::QuantizedCellParamsDynamic::linear_ih(at::Tensor const&) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xb9a6d65)
Feb 21 02:16:17     #10 0x7fd4d6fb66b6 in at::native::(anonymous namespace)::FullLayer<std::tuple<at::Tensor, at::Tensor>, at::native::(anonymous namespace)::QuantizedCellParamsDynamic>::operator()(at::Tensor const&, std::tuple<at::Tensor, at::Tensor> const&, at::native::(anonymous namespace)::QuantizedCellParamsDynamic const&) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xb9a96b6)
Feb 21 02:16:17     #11 0x7fd4d6f8642a in at::native::quantized_lstm(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xb97942a)
Feb 21 02:16:17     #12 0x7fd4d760726d in at::TypeDefault::quantized_lstm(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xbffa26d)
Feb 21 02:16:17     #13 0x7fd4dd1485b1 in torch::autograd::VariableType::(anonymous namespace)::quantized_lstm(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0x11b3b5b1)
Feb 21 02:16:17     #14 0x7fd4d773a605 in c10::detail::WrapRuntimeKernelFunctor_<std::tuple<at::Tensor, at::Tensor, at::Tensor> (*)(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool), std::tuple<at::Tensor, at::Tensor, at::Tensor>, c10::guts::typelist::typelist<at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool> >::operator()(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xc12d605)
Feb 21 02:16:17     #15 0x7fd4d773a0f5 in c10::detail::wrap_kernel_functor_unboxed_<c10::detail::WrapRuntimeKernelFunctor_<std::tuple<at::Tensor, at::Tensor, at::Tensor> (*)(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool), std::tuple<at::Tensor, at::Tensor, at::Tensor>, c10::guts::typelist::typelist<at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool> >, std::tuple<at::Tensor, at::Tensor, at::Tensor> (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool)>::call(c10::OperatorKernel*, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xc12d0f5)
Feb 21 02:16:17     #16 0x7fd4f26876f3 in std::tuple<at::Tensor, at::Tensor, at::Tensor> c10::KernelFunction::callUnboxed<std::tuple<at::Tensor, at::Tensor, at::Tensor>, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool>(c10::OperatorHandle const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x1f876f3)
Feb 21 02:16:17     #17 0x7fd4f26868dd in std::tuple<at::Tensor, at::Tensor, at::Tensor> c10::Dispatcher::callUnboxedWithDispatchKey<std::tuple<at::Tensor, at::Tensor, at::Tensor>, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool>(c10::OperatorHandle const&, c10::DispatchKey, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x1f868dd)
Feb 21 02:16:17     #18 0x7fd4f26861a7 in std::tuple<at::Tensor, at::Tensor, at::Tensor> c10::Dispatcher::callUnboxed<std::tuple<at::Tensor, at::Tensor, at::Tensor>, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool>(c10::OperatorHandle const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x1f861a7)
Feb 21 02:16:17     #19 0x7fd4f2685aed in std::tuple<at::Tensor, at::Tensor, at::Tensor> c10::OperatorHandle::callUnboxed<std::tuple<at::Tensor, at::Tensor, at::Tensor>, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool>(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool, c10::optional<c10::ScalarType>, bool) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x1f85aed)
Feb 21 02:16:17     #20 0x7fd4f2479039 in torch::autograd::THPVariable_quantized_lstm(_object*, _object*, _object*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x1d79039)
Feb 21 02:16:17     #21 0x564ab5f564f3 in _PyCFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/methodobject.c:231
Feb 21 02:16:17     #22 0x564ab5f83063 in _PyCFunction_FastCallKeywords /tmp/build/80754af9/python_1578429706181/work/Objects/methodobject.c:294
Feb 21 02:16:17     #23 0x564ab5fdddab in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4851
Feb 21 02:16:17     #24 0x564ab6001427 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3351
Feb 21 02:16:17     #25 0x564ab5fd7eba in _PyFunction_FastCall /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4933
Feb 21 02:16:17     #26 0x564ab5fd7eba in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4968
Feb 21 02:16:17     #27 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #28 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #29 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #30 0x564ab5fd80f0 in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4992
Feb 21 02:16:17     #31 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #32 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #33 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #34 0x564ab5fd87e7 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #35 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #36 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #37 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #38 0x564ab6001cf7 in do_call_core /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5120
Feb 21 02:16:17     #39 0x564ab6001cf7 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3404
Feb 21 02:16:17     #40 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #41 0x564ab5fd85cd in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #42 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #43 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #44 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #45 0x564ab5fafae0 in slot_tp_call /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:6207
Feb 21 02:16:17     #46 0x564ab5f566da in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2331
Feb 21 02:16:17     #47 0x564ab5fddefd in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4875
Feb 21 02:16:17     #48 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #49 0x564ab5fd757d in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #50 0x564ab5fd80f0 in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4992
Feb 21 02:16:17     #51 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #52 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #53 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #54 0x564ab5fd87e7 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #55 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #56 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #57 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #58 0x564ab6001cf7 in do_call_core /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5120
Feb 21 02:16:17     #59 0x564ab6001cf7 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3404
Feb 21 02:16:17     #60 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #61 0x564ab5fd85cd in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #62 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #63 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #64 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #65 0x564ab5fafae0 in slot_tp_call /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:6207
Feb 21 02:16:17     #66 0x564ab5f566da in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2331
Feb 21 02:16:17     #67 0x564ab5fddefd in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4875
Feb 21 02:16:17     #68 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #69 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #70 0x564ab5fd87e7 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #71 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #72 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #73 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #74 0x564ab6001cf7 in do_call_core /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5120
Feb 21 02:16:17     #75 0x564ab6001cf7 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3404
Feb 21 02:16:17     #76 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #77 0x564ab5fd85cd in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #78 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #79 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #80 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #81 0x564ab5fafae0 in slot_tp_call /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:6207
Feb 21 02:16:17     #82 0x564ab5f566da in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2331
Feb 21 02:16:17     #83 0x564ab5fddefd in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4875
Feb 21 02:16:17     #84 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #85 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #86 0x564ab5fd87e7 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #87 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #88 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #89 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #90 0x564ab6001cf7 in do_call_core /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5120
Feb 21 02:16:17     #91 0x564ab6001cf7 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3404
Feb 21 02:16:17     #92 0x564ab5fd7273 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #93 0x564ab5fd85cd in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #94 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #95 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #96 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #97 0x564ab5fafae0 in slot_tp_call /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:6207
Feb 21 02:16:17     #98 0x564ab5f566da in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2331
Feb 21 02:16:17     #99 0x564ab5fddefd in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4875
Feb 21 02:16:17     #100 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #101 0x564ab5fd7eba in _PyFunction_FastCall /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4933
Feb 21 02:16:17     #102 0x564ab5fd7eba in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4968
Feb 21 02:16:17     #103 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #104 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #105 0x564ab5fd7eba in _PyFunction_FastCall /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4933
Feb 21 02:16:17     #106 0x564ab5fd7eba in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4968
Feb 21 02:16:17     #107 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #108 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #109 0x564ab5fd7710 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #110 0x564ab5fd87e7 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:5084
Feb 21 02:16:17     #111 0x564ab5f568be in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2310
Feb 21 02:16:17     #112 0x564ab5f5b312 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2373
Feb 21 02:16:17     #113 0x564ab5f562fd in PyObject_Call /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2261
Feb 21 02:16:17     #114 0x564ab5faf15a in slot_tp_init /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:6420
Feb 21 02:16:17     #115 0x564ab5fde0e6 in type_call /tmp/build/80754af9/python_1578429706181/work/Objects/typeobject.c:915
Feb 21 02:16:17     #116 0x564ab5f566da in _PyObject_FastCallDict /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2331
Feb 21 02:16:17     #117 0x564ab5fd83d9 in _PyObject_FastCallKeywords /tmp/build/80754af9/python_1578429706181/work/Objects/abstract.c:2496
Feb 21 02:16:17     #118 0x564ab5fddefd in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4875
Feb 21 02:16:17     #119 0x564ab6001427 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3351
Feb 21 02:16:17     #120 0x564ab5fd757d in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #121 0x564ab5fd80f0 in fast_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4992
Feb 21 02:16:17     #122 0x564ab5fdde84 in call_function /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4872
Feb 21 02:16:17     #123 0x564ab6000669 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:3335
Feb 21 02:16:17     #124 0x564ab5fd8c08 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4166
Feb 21 02:16:17     #125 0x564ab5fd8c08 in PyEval_EvalCodeEx /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:4187
Feb 21 02:16:17     #126 0x564ab5fd99ab in PyEval_EvalCode /tmp/build/80754af9/python_1578429706181/work/Python/ceval.c:731
Feb 21 02:16:17     #127 0x564ab6059c63 in run_mod /tmp/build/80754af9/python_1578429706181/work/Python/pythonrun.c:1025
Feb 21 02:16:17     #128 0x564ab605a060 in PyRun_FileExFlags /tmp/build/80754af9/python_1578429706181/work/Python/pythonrun.c:978
Feb 21 02:16:17     #129 0x564ab605a262 in PyRun_SimpleFileExFlags /tmp/build/80754af9/python_1578429706181/work/Python/pythonrun.c:419
Feb 21 02:16:17     #130 0x564ab605dd52 in run_file /tmp/build/80754af9/python_1578429706181/work/Modules/main.c:340
Feb 21 02:16:17     #131 0x564ab605dd52 in Py_Main /tmp/build/80754af9/python_1578429706181/work/Modules/main.c:811
Feb 21 02:16:17     #132 0x564ab5f283bd in main /tmp/build/80754af9/python_1578429706181/work/Programs/python.c:69
Feb 21 02:16:17     #133 0x7fd4f943282f in __libc_start_main /build/glibc-LK5gWL/glibc-2.23/csu/../csu/libc-start.c:291
Feb 21 02:16:17     #134 0x564ab6007083 in _start /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:103
Feb 21 02:16:17 
Feb 21 02:16:17 SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/caffe2/utils/threadpool/pthreadpool.cc:332:13 in 
Feb 21 02:16:17 Traceback (most recent call last):
Feb 21 02:16:17   File "test/run_test.py", line 487, in <module>
Feb 21 02:16:17     main()
Feb 21 02:16:17   File "test/run_test.py", line 480, in main
Feb 21 02:16:17     raise RuntimeError(message)
Feb 21 02:16:17 RuntimeError: test_quantization failed!
Feb 21 02:16:17 + cleanup
```

Could you take a look?

supriyar commented:
@jianyuh that's strange. I ran this locally and it works on my machine. It looks like mismatch in function signature.
Can you try changing the function arguments [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/fc-dynamic-run.cc#L22) to remove the `const` qualifier and try?

jianyuh committed:
Update on "[pt][quant] Enable the dynamic quant LSTM unit test by removing hypothesis"


Check more discussions in https://github.com/pytorch/pytorch/issues/32644.

Previously, after landing https://github.com/pytorch/pytorch/pull/32757, we found the flaky tests on `test_quantized_rnn` in some specific CI environment, where we cannot reproduce in our devserver.

The error is reported as it failed the `scale > 0` check in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L69 .

- Step 1: we disabled the unit tests in https://github.com/pytorch/pytorch/pull/32742

- Step 2: we investigated the FBGEMM `ChooseQuantizationParams` routine and added assert for min, max, qmin, qmax, but we didn't find the abnormal numbers :
https://github.com/pytorch/pytorch/pull/32739

scale is calculated as the following:
float scale = (static_cast<double>(max) - min) / (qmax - qmin);

If max is larger than min, and qmax is larger than qmin, then scale should be always positive.

- Step 3: we tried to print the intermediate result of x_min and x_max in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L59-L63. Check https://github.com/pytorch/pytorch/pull/33167 for the places we add the prints.

One interesting test case caused the failure:
```
Jan 29 23:46:57 x_min: -nan, x_max: -nan
Jan 29 23:46:57 input: -nan 0.195275 -0.380797 -0.510193 -0.662963 0.761594 0.923773 -nan -0.123308 -0.380797 -0.761594 -0.603171 0.906278 0.000028 -nan 0.462941 -0.380797 -0.651336 0.717926 0.761594 0.460751 -nan 0.945544 -0.380797 -0.334840 0.831535 0.761594 0.793159 -nan -0.752703 -0.380797 -0.761594 0.809177 0.761594 0.955718 -nan -0.717292 -0.380797 -0.650404 0.790869 0.761594 0.812008 -nan 0.917088 -0.380797 -0.761594 0.829536 0.761594 0.000000 -nan 0.855218 -0.380796 -0.036113 0.815578 0.761594 0.942870 -nan -0.675926 -0.380797 -0.755426 -0.658459 0.000000 0.000000 -nan 0.490974 -0.380797 -0.725917 -0.539679 0.761594 0.836110 -nan -0.714314 -0.380797 -0.761594 -0.738534 0.761594 0.791485 -nan 0.345541 -0.380797 -0.670953 -0.609272 0.962786 0.834005 -nan -0.583866 -0.380797 -0.761594 -0.620485 0.750284 0.854029 -nan -0.661655 -0.380797 -0.282387 0.863691 0.761593 0.956279 -nan -0.667277 -0.380797 -0.730052 -0.732839 0.761594 0.002185 -nan 0.380570 -0.380797 -0.453875 -0.523069 0.834210 0.792132
```

More detailed test results are here:
https://circleci.com/gh/pytorch/pytorch/4374802?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Basically, x_min and x_max are both -nan, which failed the check for x_min <= x_max. Actually if any operands for a comparison is nan (e.g., nan < 3, nan < nan, nan <= nan), the result for the comparison will always be False. The culprit should be this NaN value.

The issue is where we generated such NaN value?

- Step 4:
We analyzed that NaN can possibly come from two sources:

    - Unit test: when when we initialize the `torch.Tensor` or `torch.rand` somewhere, it generates NaN number;

    - Somewhere in `aten/src/ATen/native/RNN.cpp` or `torch/nn/quantized/dynamic/modules/rnn.py`, we used some illegal instructions.

As there is no such issue before https://github.com/pytorch/pytorch/pull/32757, we suspect some changes in that PR triggered this issue.

In https://github.com/pytorch/pytorch/pull/33167, we avoided using hypothesis by removing `@given(qengine=st.sampled_from(("qnnpack", "fbgemm"))):` and all tests pass.

It looks like that the hypothesis added in https://github.com/pytorch/pytorch/pull/32757 triggered the generation of NaN in some input tensors.


- Step 5:

We still need to find the root cause for NaN, either it is the problem of hypothesis itself, or hypothesis generated more random cases which helped triggered some NaN numbers in `RNN.cpp` or `rnn.py`.

Differential Revision: [D20016038](https://our.internmc.facebook.com/intern/diff/D20016038/)

[ghstack-poisoned]

jianyuh commented:
@supriyar Thanks for the quick reply! Re-tried with the new function signature with `const` removed.

Regarding that error, https://github.com/pytorch/pytorch/pull/33167 doesn't seem to have such issue. The difference is that this PR also tests `qnnpack` engine.

jianyuh committed:
Update on "[pt][quant] Enable the dynamic quant LSTM unit test by removing hypothesis"


Check more discussions in https://github.com/pytorch/pytorch/issues/32644.

Previously, after landing https://github.com/pytorch/pytorch/pull/32757, we found the flaky tests on `test_quantized_rnn` in some specific CI environment, where we cannot reproduce in our devserver.

The error is reported as it failed the `scale > 0` check in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L69 .

- Step 1: we disabled the unit tests in https://github.com/pytorch/pytorch/pull/32742

- Step 2: we investigated the FBGEMM `ChooseQuantizationParams` routine and added assert for min, max, qmin, qmax, but we didn't find the abnormal numbers :
https://github.com/pytorch/pytorch/pull/32739

scale is calculated as the following:
float scale = (static_cast<double>(max) - min) / (qmax - qmin);

If max is larger than min, and qmax is larger than qmin, then scale should be always positive.

- Step 3: we tried to print the intermediate result of x_min and x_max in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L59-L63. Check https://github.com/pytorch/pytorch/pull/33167 for the places we add the prints.

One interesting test case caused the failure:
```
Jan 29 23:46:57 x_min: -nan, x_max: -nan
Jan 29 23:46:57 input: -nan 0.195275 -0.380797 -0.510193 -0.662963 0.761594 0.923773 -nan -0.123308 -0.380797 -0.761594 -0.603171 0.906278 0.000028 -nan 0.462941 -0.380797 -0.651336 0.717926 0.761594 0.460751 -nan 0.945544 -0.380797 -0.334840 0.831535 0.761594 0.793159 -nan -0.752703 -0.380797 -0.761594 0.809177 0.761594 0.955718 -nan -0.717292 -0.380797 -0.650404 0.790869 0.761594 0.812008 -nan 0.917088 -0.380797 -0.761594 0.829536 0.761594 0.000000 -nan 0.855218 -0.380796 -0.036113 0.815578 0.761594 0.942870 -nan -0.675926 -0.380797 -0.755426 -0.658459 0.000000 0.000000 -nan 0.490974 -0.380797 -0.725917 -0.539679 0.761594 0.836110 -nan -0.714314 -0.380797 -0.761594 -0.738534 0.761594 0.791485 -nan 0.345541 -0.380797 -0.670953 -0.609272 0.962786 0.834005 -nan -0.583866 -0.380797 -0.761594 -0.620485 0.750284 0.854029 -nan -0.661655 -0.380797 -0.282387 0.863691 0.761593 0.956279 -nan -0.667277 -0.380797 -0.730052 -0.732839 0.761594 0.002185 -nan 0.380570 -0.380797 -0.453875 -0.523069 0.834210 0.792132
```

More detailed test results are here:
https://circleci.com/gh/pytorch/pytorch/4374802?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Basically, x_min and x_max are both -nan, which failed the check for x_min <= x_max. Actually if any operands for a comparison is nan (e.g., nan < 3, nan < nan, nan <= nan), the result for the comparison will always be False. The culprit should be this NaN value.

The issue is where we generated such NaN value?

- Step 4:
We analyzed that NaN can possibly come from two sources:

    - Unit test: when when we initialize the `torch.Tensor` or `torch.rand` somewhere, it generates NaN number;

    - Somewhere in `aten/src/ATen/native/RNN.cpp` or `torch/nn/quantized/dynamic/modules/rnn.py`, we used some illegal instructions.

As there is no such issue before https://github.com/pytorch/pytorch/pull/32757, we suspect some changes in that PR triggered this issue.

In https://github.com/pytorch/pytorch/pull/33167, we avoided using hypothesis by removing `@given(qengine=st.sampled_from(("qnnpack", "fbgemm"))):` and all tests pass.

It looks like that the hypothesis added in https://github.com/pytorch/pytorch/pull/32757 triggered the generation of NaN in some input tensors.


- Step 5:

We still need to find the root cause for NaN, either it is the problem of hypothesis itself, or hypothesis generated more random cases which helped triggered some NaN numbers in `RNN.cpp` or `rnn.py`.

Differential Revision: [D20016038](https://our.internmc.facebook.com/intern/diff/D20016038/)

[ghstack-poisoned]

jianyuh commented:
Update: "fbgemm" qengine can pass the unit tests: #33621
